### PR TITLE
Parse jediyaml only once

### PR DESCRIPTION
**Pull request by aerorahul**
# Description
`JEDIYAML` was being parsed 3 times; once in `get_obs_dict`, second in `get_bias_dict` and a third time in `initialize` for the specific component analysis task.

This PR:
- eliminates the duplications and constructs the `jedi_config` dictionary just once.  The dictionary is written out before calling the executable.
- updates hash to gdasapp
- updates configs for snow, aerosol, atmvar and atmens JEDI-DA to include `JEDI_FIX_YAML` and `CRTM_FIX_YAML` .  This allows greater flexibility and control over the contents of these fix data sets to be copied into the run directory.
- Combines snowDA and aerosolDA into a single test (Thank you @CoryMartin-NOAA; #2323)

# Type of change

- Maintenance (removes duplication and potential to make a mistake)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO

# How has this been tested?
- snow + aerosol DA test on Hera was successful

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: ush/python/pygfs/task/analysis.py
Comment (CoryMartin-NOAA) [inline ush/python/pygfs/task/analysis.py]:
```suggestion
        This method extracts 'observers' from the JEDI yaml and from that list, extracts a list of
```

File: ush/python/pygfs/task/analysis.py
Comment (CoryMartin-NOAA) [inline ush/python/pygfs/task/analysis.py]:
```suggestion
        This method extracts 'observers' from the JEDI yaml and from that list, extracts a list of
```

File: ush/python/pygfs/task/analysis.py
Comment (CoryMartin-NOAA) [inline ush/python/pygfs/task/analysis.py]: this is nice!

**Review by CoryMartin-NOAA — APPROVED** [review]
Approve, but suggested a couple of minor grammatical things in comments. Thanks!

**Review by jiaruidong2017 — APPROVED** [review]
It looks good to me.

**Review by CoryMartin-NOAA — APPROVED** [review]

File: ush/python/pygfs/task/analysis.py
Comment (aerorahul) [inline ush/python/pygfs/task/analysis.py]:
```suggestion
        from the component directory.
        TODO: COM_ATMOS_ANALYSIS_PREV is hardwired here and this method is not appropriate in `analysis.py` and should be implemented in the component where this is applicable.
```
@RussTreadon-NOAA We should implement this in `atm_analysis.py`

File: ush/python/pygfs/task/analysis.py
Comment (aerorahul) [inline ush/python/pygfs/task/analysis.py]:
```suggestion
        TODO: COM_ATMOS_ANALYSIS_PREV is hardwired here and this method is not appropriate in
        `analysis.py` and should be implemented in the component where this is applicable.
```

File: ush/python/pygfs/task/analysis.py
Comment (DavidHuber-NOAA) [inline ush/python/pygfs/task/analysis.py]: Seems like this should return an error.  Since your recursive calls below are checking for `dict` instances, this would only return `None` if the parent `nested_dict` were not a `dict`.  Making this an error would make debugging easier.

File: ush/python/pygfs/task/analysis.py
Comment (RussTreadon-NOAA) [reply ush/python/pygfs/task/analysis.py]: GDAS (and GFS) ATM and EnKF analysis use radiance (and later aircraft) bias correction files from the GDAS.  Do we replicate `get_bias_dict` `atm_analysis.py` and `atmens_analysis.py`?  Other JEDI DA modes (land, ocean) may also use GDAS bias correction files in the future.  Where's the best place to put a staging method which copies files from the GDAS to various JEDI DA modes?

**Review by DavidHuber-NOAA — CHANGES_REQUESTED** [review]
One suggestion regarding error handling in the new `find_value_in_nested_dict` function.  Otherwise looks really nice!

File: ush/python/pygfs/task/analysis.py
Comment (aerorahul) [reply ush/python/pygfs/task/analysis.py]: [e2c9a51](https://github.com/NOAA-EMC/global-workflow/pull/2387/commits/e2c9a512444ff805fd31252b244b42a0a6cc4200) update will ensure a `KeyError` is thrown when the key is not found.
Additionally, a `TypeError` is caught if the incoming data is not a `dict`.

**Review by DavidHuber-NOAA — APPROVED** [review]
Looks good, thanks!

File: ush/python/pygfs/task/analysis.py
Comment (WalterKolczynski-NOAA) [inline ush/python/pygfs/task/analysis.py]: This part of the docstring lost its indentation.

File: ush/python/pygfs/task/analysis.py
Comment (WalterKolczynski-NOAA) [inline ush/python/pygfs/task/analysis.py]: As written, this is going to have strange (possible inconsistent) behavior for keys that appear multiple times. If the top-most dict contains the target key, that will be the value returned. Otherwise, it will iterate through keys looking for the target key, but dicts should not be considered naturally ordered (even though the current implementation of python preserves ordering), so if the target key is contained in multiple nested dicts (but not the top one), the value retrieved could change from execution to execution.

Even without the undetermined ordering issue, the retrieved value won't necessarily be the shallowest (unless it is in the top-level dict) or the deepest matching key.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: ush/python/pygfs/task/analysis.py
Comment (aerorahul) [reply ush/python/pygfs/task/analysis.py]: This is true, and thus, I don't think we can move it to a generic module.
Think of giving someone just a house number to find and leaving them at the edge of town.
Which specific house they find will depend on the choices they make; left, or right.
Same is with this search.  This will return the first key the search encounters.

Perhaps we should remove this method and instead look for `observations` under `cost function`.  If it finds it, then thats it.  If not, it should look for `observations` at the top-level. 
Those are the use cases for now.
If we run into problems, we can revisit this and implement a `find_observations` method targeted for this rather than a generic `find_key_in_nested_dict`.

File: ush/python/pygfs/task/analysis.py
Comment (aerorahul) [inline ush/python/pygfs/task/analysis.py]:
```suggestion
    nested_dict = {
        'a': {
            'b': {
                'c': 1,
                'd': {
                    'e': 2,
                    'f': 3
                }
            },
            'g': 4
        },
        'h': {
            'i': 5
        },
        'j': {
            'k': 6
        }
    }

    user_key = input("Enter the key to search for: ")
    result = find_value_in_nested_dict(nested_dict, user_key)
```

**Review by WalterKolczynski-NOAA — APPROVED** [review]

File: ush/forecast_predet.sh
Comment (WalterKolczynski-NOAA) [reply ush/forecast_predet.sh:262]: Since this script is always sourced, can we try putting
```
# shellcheck disable=SC2034
```
at the top of the script (directly after the shebang) and remove the other instances?

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

**Review by WalterKolczynski-NOAA — COMMENTED** [review]
Looks good so far. Holding approval until marked ready.

**Comment by CoryMartin-NOAA** [comment]
Just needs this commit hash: https://github.com/NOAA-EMC/GDASApp/commit/82e16bbf9eafec67e517827357ab8e1a360740fc

**Comment by aerorahul** [comment]
> Just needs this commit hash: [NOAA-EMC/GDASApp@82e16bb](https://github.com/NOAA-EMC/GDASApp/commit/82e16bbf9eafec67e517827357ab8e1a360740fc)

Thanks @CoryMartin-NOAA 
I am running a cycled test on Hera to confirm everything works with the `develop` of GDASApp and will mark this PR ready when that completes.

File: scripts/exglobal_oceanice_products.py
Comment (DavidHuber-NOAA) [inline scripts/exglobal_oceanice_products.py]: Same as #2398:

```suggestion
            f'COM_{ocnice.task_config.component.upper()}_HISTORY', f'COM_{ocnice.task_config.component.upper()}_GRIB',
```

**Review by DavidHuber-NOAA — CHANGES_REQUESTED** [review]

**Review by WalterKolczynski-NOAA — APPROVED** [review]

File: parm/gdas/aero_crtm_coeff.yaml.j2
Comment (DavidHuber-NOAA) [inline parm/gdas/aero_crtm_coeff.yaml.j2:1]: Is there a reason this yaml was written new instead of moving/editing `aero_ctrm_coeff.yaml`?

**Review by CoryMartin-NOAA — APPROVED** [review]
Looks good, thanks!

**Review by DavidHuber-NOAA — COMMENTED** [review]
Just one question regarding one of the yamls.  I'd prefer to be able to keep history tracking wherever possible.

File: parm/gdas/aero_crtm_coeff.yaml.j2
Comment (aerorahul) [reply parm/gdas/aero_crtm_coeff.yaml.j2:1]: I did a `git mv aero_ctrm_coeff.yaml aero_ctrm_coeff.yaml.j2`.
However, somehow github thinks it is new instead of a rename.
Perhaps because, in addition to moving, I edited the contents of the file.
🤷🏾

**Comment by aerorahul** [comment]
@CoryMartin-NOAA 
Sadly the new test did not run in the CI because Jenkins currently does not know how to dynamically add new tests.
An update to the CI Jenkinsfile is needed.  I will push that update shortly.

File: parm/gdas/aero_crtm_coeff.yaml.j2
Comment (DavidHuber-NOAA) [reply parm/gdas/aero_crtm_coeff.yaml.j2:1]: Okie dokie, it is what it is.

**Review by DavidHuber-NOAA — APPROVED** [review]
Approve pending completion of the updated CI test.

**Comment by aerorahul** [comment]
The atm+aero+snow DA test is now triggered with the updated Jenkinsfile.
Screenshot 2024-03-15 at 2 23 42 PM

**Comment by WalterKolczynski-NOAA** [comment]
```
[38;5;39m2024-03-16 01:22:21,609 - DEBUG    - aero_analysis: ( <pygfs.task.aero_analysis.AerosolAnalysis object at 0x2af7d4fa6590> )[0m
[38;21m2024-03-16 01:22:21,615 - INFO     - analysis    : BEGIN: pygfs.task.analysis.add_fv3_increments[0m
[38;5;39m2024-03-16 01:22:21,616 - DEBUG    - analysis    : ( <pygfs.task.aero_analysis.AerosolAnalysis object at 0x2af7d4fa6590>, '/scratch1/NCEPDEV/stmp2/Terry.McGuinness/RUNDIRS/C96_atmaerosnowDA_2d816209/gdasaeroanl_00/anl/aeroinc.20211221.000000.fv_tracer.res.tile{tilenum}.nc', '/scratch1/NCEPDEV/global/CI/2387/RUNTESTS/COMROOT/C96_atmaerosnowDA_2d816209/gdas.20211220/18//model_data/atmos/restart/20211221.000000.fv_tracer.res.tile{tilenum}.nc', ['dust1', 'dust2', 'dust3', 'dust4', 'dust5', 'seas1', 'seas2', 'seas3', 'seas4', 'so4', 'oc1', 'oc2', 'bc1', 'bc2'] )[0m
HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 47243674934976:
  #000: H5F.c line 509 in H5Fopen(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #001: H5Fint.c line 1400 in H5F__open(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #002: H5Fint.c line 1615 in H5F_open(): unable to lock the file
    major: File accessibilty
    minor: Unable to open file
  #003: H5FD.c line 1640 in H5FD_lock(): driver lock request failed
    major: Virtual File Layer
    minor: Can't update object
  #004: H5FDsec2.c line 941 in H5FD_sec2_lock(): unable to lock file, errno = 11, error message = 'Resource temporarily unavailable'
    major: File accessibilty
    minor: Bad file ID accessed
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/CI/2387/gfs/scripts/exglobal_aero_analysis_finalize.py", line 24, in <module>
    AeroAnl.finalize()
  File "/scratch1/NCEPDEV/global/CI/2387/gfs/ush/python/wxflow/logger.py", line 266, in wrapper
    retval = func(*args, **kwargs)
  File "/scratch1/NCEPDEV/global/CI/2387/gfs/ush/python/pygfs/task/aero_analysis.py", line 182, in finalize
    self._add_fms_cube_sphere_increments()
  File "/scratch1/NCEPDEV/global/CI/2387/gfs/ush/python/wxflow/logger.py", line 266, in wrapper
    retval = func(*args, **kwargs)
  File "/scratch1/NCEPDEV/global/CI/2387/gfs/ush/python/pygfs/task/aero_analysis.py", line 214, in _add_fms_cube_sphere_increments
    super().add_fv3_increments(inc_template, bkg_template, incvars)
  File "/scratch1/NCEPDEV/global/CI/2387/gfs/ush/python/wxflow/logger.py", line 266, in wrapper
    retval = func(*args, **kwargs)
  File "/scratch1/NCEPDEV/global/CI/2387/gfs/ush/python/pygfs/task/analysis.py", line 159, in add_fv3_increments
    with Dataset(inc_path, mode='r') as incfile, Dataset(bkg_path, mode='a') as rstfile:
  File "netCDF4/_netCDF4.pyx", line 2135, in netCDF4._netCDF4.Dataset.__init__
  File "netCDF4/_netCDF4.pyx", line 1752, in netCDF4._netCDF4._ensure_nc_success
OSError: [Errno -101] NetCDF: HDF error: b'/scratch1/NCEPDEV/global/CI/2387/RUNTESTS/COMROOT/C96_atmaerosnowDA_2d816209/gdas.20211220/18//model_data/atmos/restart/20211221.000000.fv_tracer.res.tile4.nc'
+ JGLOBAL_AERO_ANALYSIS_FINALIZE[1]: postamble JGLOBAL_AERO_ANALYSIS_FINALIZE 1710552131 1
+ preamble.sh[70]: set +x
End JGLOBAL_AERO_ANALYSIS_FINALIZE at 01:22:24 with error code 1 (time elapsed: 00:00:13)
```

**Comment by aerorahul** [comment]
Interesting. I did not encounter this error when I was testing outside the CI.

**Comment by TerrenceMcGuinness-NOAA** [comment]
@aerorahul Maybe **tilenum** is not resolving in [ero_analysis.py
](https://github.com/aerorahul/global-workflow/blob/3420ed3a1ef02010606498833fcad5bae58f62ad/ush/python/pygfs/task/aero_analysis.py#L171)

Or is this the correct file name with said **`tilenum=4`** that failed to read:

```
Terry.McGuinness (hfe08) gfs ((HEAD detached at 2d816209)) $ ls -l /scratch1/NCEPDEV/global/CI/2387/RUNTESTS/COMROOT/C96_atmaerosnowDA_2d816209/gdas.20211220/18//model_data/atmos/restart/20211221.000000.fv_tracer.res.tile4.nc
-rw-r--r-- 1 Terry.McGuinness stmp 145624564 Mar 16 01:16 /scratch1/NCEPDEV/global/CI/2387/RUNTESTS/COMROOT/C96_atmaerosnowDA_2d816209/gdas.20211220/18//model_data/atmos/restart/20211221.000000.fv_tracer.res.tile4.nc
```

**Comment by aerorahul** [comment]
I ran a test over the weekend on Hera.  
I did not encounter the issue in [this comment](https://github.com/NOAA-EMC/global-workflow/pull/2387#issuecomment-2001098958)

The experiment and COM directories are at:
```
/scratch1/NCEPDEV/stmp2/Rahul.Mahajan/RUNTESTS/EXPDIR/aas
/scratch1/NCEPDEV/stmp2/Rahul.Mahajan/RUNTESTS/COMROOT/aas
```

**Comment by aerorahul** [comment]
Screenshot 2024-03-18 at 9 29 23 AM
Screenshot 2024-03-18 at 9 29 43 AM

**Comment by TerrenceMcGuinness-NOAA** [comment]
```
Task: gdasaeroanlfinal
  account: nems
  command: /scratch1/NCEPDEV/global/CI/2387/gfs/jobs/rocoto/aeroanlfinal.sh
  cores: 40
  cycledefs: gdas
  final: false
  jobname: C96_atmaerosnowDA_2d816209_gdasaeroanlfinal_00
  join: /scratch1/NCEPDEV/global/CI/2387/RUNTESTS/COMROOT/C96_atmaerosnowDA_2d816209/logs/2021122100/gdasaeroanlfinal.log
  maxtries: 2
  memory: 3072M
  name: gdasaeroanlfinal
  nodes: 1:ppn=40:tpp=1
  partition: hera
  queue: batch
  throttle: 9999999
  walltime: 00:10:00
  environment
    CDATE ==> 2021122100
    CDUMP ==> gdas
    COMROOT ==> /scratch1/NCEPDEV/global/CI/2387/RUNTESTS/COMROOT
    DATAROOT ==> /scratch1/NCEPDEV/stmp2/Terry.McGuinness/RUNDIRS/C96_atmaerosnowDA_2d816209
    EXPDIR ==> /scratch1/NCEPDEV/global/CI/2387/RUNTESTS/EXPDIR/C96_atmaerosnowDA_2d816209
    HOMEgfs ==> /scratch1/NCEPDEV/global/CI/2387/gfs
    NET ==> gfs
    PDY ==> 20211221
    RUN ==> gdas
    RUN_ENVIR ==> emc
    cyc ==> 00
  dependencies
    AND is satisfied
      gdasaeroanlrun of cycle 202112210000 is SUCCEEDED

Cycle: 202112210000
  Valid for this task: YES
  State: active
  Activated: 2024-03-15 18:08:11 UTC
  Completed: -
  Expired: -

Job: 57318802
  State:  SUCCEEDED (COMPLETED)
  Exit Status:  0
  Tries:  1
  Unknown count:  0
```

@aerorahul I just did that now and **gdasaeroanlfinal** succeeded from the `EXPDIR` in place that Jenkins CI had placed on disk.

**Comment by TerrenceMcGuinness-NOAA** [comment]
I just did that now and **gdasaeroanlfinal** from the **EXPDIR** succeed in place from a rewind that Jenkins CI had placed on disk and was about to mention I was going to restart Jenkins and it would have labeled this PR passed in just a few minutes.

**Comment by RussTreadon-NOAA** [comment]
I'm getting [reports](https://github.com/NOAA-EMC/JEDI-T2O/pull/111#issuecomment-2004385016) of JEDI ATM job `gdastmanlinit` failing after this PR was merged into develop.

**Comment by aerorahul** [comment]
Are the reporters using the `configs` from the global-workflow or something else?

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA identified the problem.   This PR added `CRTM_FIX_YAML` and `JEDI_FIX_YAML` to `config.atmanl`.

**Comment by RussTreadon-NOAA** [comment]
DA developers working on GDAS-validation use a JEDI-T2O front end with a modified `config.atmanl`.  This `config.atmanl` needed to be updated as Cory pointed out.

**Comment by aerorahul** [comment]
Thanks @RussTreadon-NOAA 
I was not aware of the modified `config.atmanl` in a different location.

**Pull request merged**
